cpp/src/groupby/hash/groupby.cu

```diff
@@ -118,74 +120,135 @@ bool constexpr is_hash_aggregation(aggregation::Kind t)
   return array_contains(hash_aggregations, t);
 }
 
+bool is_hash_aggregation(aggregation const& agg)
+{
+  if (agg.kind == aggregation::NTH_ELEMENT) {
+    auto const& n = dynamic_cast<cudf::detail::nth_element_aggregation const&>(agg)._n;
+    return n == 0 || n == -1;  // Only `FIRST` and `LAST` are supported for now.
+  }
+  return is_hash_aggregation(agg.kind);
+}
+
 class groupby_simple_aggregations_collector final
   : public cudf::detail::simple_aggregations_collector {
  public:
+  groupby_simple_aggregations_collector(column_view original_agg_column,
+                                        rmm::cuda_stream_view stream)
+    : original_agg_column{original_agg_column}, stream{stream}
+  {
+  }
+
   using cudf::detail::simple_aggregations_collector::visit;
 
+  // Default aggregation collection.
+  std::vector<std::unique_ptr<aggregation>> visit(data_type, aggregation const& agg) override
+  {
+    aggs_and_column_views.push_back(std::pair(agg.clone(), original_agg_column));
+    return {};
```

Comment:
I'd be delighted to discuss this at your earliest convenience.

The `cudf::detail::simple_aggregations_collector` interface is used (in several places, including here) to convert compound aggregations like `MEAN` into simpler constituents (`SUM`, `COUNT`), via a `visitor`. Note that `cudf::aggregation` is aware of `simple_aggregations_collector`, for the visitor pattern to work.

Unfortunately, the collector only collects the new *aggregations*. It did not foresee a case where the simpler aggregations might be applied to a new, temporary column instead of the original aggregation column.

E.g. `NTH_ELEMENT(0)` is implemented using a groupby `MIN` on a temporarily generated `source-index` column:
```
agg_column : [ a, b, c, d, e, f, g, h ]
gby_column : [ 0, 0, 0, 0, 1, 1, 1, 1 ]
idx_column : [ 0, 1, 2, 3, 4, 5, 6, 7 ] //<--- Temporary, generated column.
min(idx)   : [ 0,          4          ]
gather(min): [ a,          e          ]
```

For this to work via `simple_aggregations_collector`, `NTH_ELEMENT(0)` should map to both `MIN` *and* the `idx_column`. But `simple_aggregations_collector::visit()` returns only `vector<unique_ptr<aggregation>>`.

The minimal change to support this sort of collection involved simply making the `collector` stateful, and storing the mapping in the `collector`. To continue to conform to the `simple_aggregations_collector` interface, I'm compelled to return a `vector`.

Reply:
Note, further, that instead of having `flatten_simple_aggregations` return an increasingly wider `tuple`, I have used a class to accumulate the results.

Reply:
My suggestion is, Don't modify `groupby_simple_aggregations_collector` and its return values; Modify `insert_agg` lambda to add special case for nth element values, and make `flatten_single_pass_aggs` also to return additional owning vector of columns created in `insert_agg`.
The change will be simpler. Since `insert_agg` is already doing the work of avoiding duplicate aggregations, it can be modified, or take additional responsibility of special case of nth element (first, last), or add extra lambda function to take this responsibility.

Reply:
Following up with notes from a call with @mythrocks earlier today. We discussed this design at length, and proposed a number of alternatives. We concluded that this design is undesireable for a few reasons. In addition to the unused return values, it makes the collector _stateful_ which isn't true for other machinery derived from the same base class. There's also serious questions about managing the lifetimes of the extra temporary columns and how we would be able to indicate to the finalizer that an `NTH_ELEMENT` aggregation request for a column view depends on a `MIN` aggregation over a _different_ column view (the temporary column whose lifetime would be managed externally). We agreed it would be best to avoid these pitfalls.

We decided on a pretty substantial rewrite. This will involve the following steps:

1. Add new aggregations `MIN_INDEX_VALID`, `MIN_INDEX_ALL`, `MAX_INDEX_VALID`, `MAX_INDEX_ALL`. These will be hash-based aggregations that return the min/max index of (non-null / any) value in the group. They will be implemented [similarly to `ARGMIN` and `ARGMAX`](https://github.com/rapidsai/cudf/blob/48dc1681da8a6f532e664d740a187ede6885968b/cpp/include/cudf/detail/aggregation/aggregation.cuh#L463-L488), but won't compare values -- only source indices, and ignoring null values depending on which variation is selected. The target type is always `size_type`, [just like for `ARGMIN`](https://github.com/rapidsai/cudf/blob/e099e01c9b6ab8a2db5d5ee446b8843ee6199acc/cpp/include/cudf/detail/aggregation/aggregation.hpp#L1252-L1256). At first, these will _only_ support hash-based groupby (a sort-based implementation is possible using the same machinery as `ARGMIN` and `ARGMAX` with slight alterations but is out of scope for the initial pass).
2. Update the [`can_use_hash_groupby`](https://github.com/rapidsai/cudf/blob/48dc1681da8a6f532e664d740a187ede6885968b/cpp/src/groupby/hash/groupby.cu#L630) to raise an error if a sort groupby is required by some aggregation request and a `[MIN/MAX]_INDEX_[VALID/ALL]` is manually requested (not as part of an `NTH_ELEMENT` request).
3. Implement a hash-based `NTH_ELEMENT` that requests a "simple" aggregation for one of `[MIN/MAX]_INDEX_[VALID/ALL]` and then performs a gather. This will look like [how we currently compute `MIN` as a hash-based aggregation for strings](https://github.com/rapidsai/cudf/blob/48dc1681da8a6f532e664d740a187ede6885968b/cpp/src/groupby/hash/groupby.cu#L274). The benefit here is that the `MIN_INDEX_VALID` (and similar) aggregations apply to the _source column_ and require no additional column allocation or lifetime extension to work correctly in the results cache.
4. In the future, add a sort-based implementation for the index aggregations if needed.

Reply:
IMHO, this does not scale really well. 4 new aggregations for 4 scenarios of single NTH_ELEMENT aggregation! 
It would be nice to analyse this again.

Reply:
@karthikeyann I'd be happy to discuss -- the previous approach of this PR required drastic changes to the internals of how aggregations are collected and finalized, while the introduction of `[MIN/MAX]_INDEX_[VALID/ALL]` (or some parameterized aggregation containing that behavior) allows us to leverage the existing machinery and no significant alterations to the framework.